verilog/analysis/checkers/forbid_implicit_declarations_rule_test.cc

```diff
@@ -0,0 +1,245 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/forbid_implicit_declarations_rule.h"
+
+#include <initializer_list>
+
+#include "gtest/gtest.h"
+#include "common/analysis/linter_test_utils.h"
+#include "common/analysis/text_structure_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/CST/verilog_treebuilder_utils.h"
+#include "verilog/analysis/verilog_analyzer.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(ForbidImplicitDeclarationsRule, FunctionFailures) {
+  auto kToken = SymbolIdentifier;
+  const std::initializer_list<LintTestCase> ForbidImplicitDeclarationsTestCases = {
+      {""},
+      {"module m;\nendmodule\n"},
+      {"module m;\nassign ", {kToken, "a1"}, " = 1'b0;\nendmodule"},
+      {"module m;\n"
+       "  wire a1;\n"
+       "  assign a1 = 1'b0;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "  module foo;\n"
+       "  endmodule\n"
+       "endmodule"},
+      {"module m;\n"
+       "  module foo;\n"
+       "  endmodule\n"
+       "  assign ", {kToken, "a1"}, " = 1'b1;\n"
+       "endmodule"},
+      {"module m;\n"
+       "  wire a1;\n"
+       "  module foo;\n"
+       "    assign a1 = 1'b0;\n"
```

Comment:
Variable a1 is not visible in module foo, linter should fail here

Reply:
According to LRM 23.4 `Nested modules`:
```
A module can be declared within another module. The outer name space is visible
to the inner module so that any name declared there can be used
```

example (modelsim):
```
  module m;
    wire a1;
    assign a1 = 1'b0;

    module foo;
      assign a1 = 1'b1;
    endmodule;

    initial $monitor("a1 = %b", a1);
  endmodule
```
output is an invalid 'X' value:
```
# a1 = x
```

Reply:
Argh, you are right, sorry. SystemVerilog is so ugly sometimes. I was used to nested modules with ports, which needs to be instantiated manually and not implicitly like your example.

I didn't know they had access to the parent namespace.

Reply:
Thanks both of you for your scrutiny over the test cases.

I also just learned this about implicit instantiation (LRM 23.4):

_"Nested modules with no ports that are not explicitly instantiated shall be implicitly instantiated once with an instance name identical to the module name. Otherwise, if they have ports and are not explicitly instantiated, they are ignored."_